### PR TITLE
Fix organize imports overlap

### LINKED ISSUE #43107: Overlapping changes error for organize imports
# Bug Report

From https://github.com/microsoft/vscode/issues/118098

### 🔎 Search Terms

- organize imports

### 🕗 Version & Regression Information
4.3.0-dev.20210303

### 💻 Code

Run organize imports on the following code:

```ts
import * as something from "path";/** 
 * some comment here
 * and there
 */
import * as somethingElse from "anotherpath";
```

### 🙁Bug

```
[Trace  - 21:25:10.955] <semantic> Response received: organizeImports (8). Request took 1 ms. Success: false . Message: Error processing request. Debug Failure. False expression: Changes overlap
Verbose Debug Information: {"pos":0,"end":77} and {"pos":39,"end":122}
Error: Debug Failure. False expression: Changes overlap
Verbose Debug Information: {"pos":0,"end":77} and {"pos":39,"end":122}
    at _loop_10 (/Users/matb/projects/san/node_modules/typescript/lib/tsserver.js:136657:34)
    at /Users/matb/projects/san/node_modules/typescript/lib/tsserver.js:136663:25
    at Object.mapDefined (/Users/matb/projects/san/node_modules/typescript/lib/tsserver.js:612:30)
    at Object.getTextChangesFromChanges (/Users/matb/projects/san/node_modules/typescript/lib/tsserver.js:136651:27)
    at ChangeTracker.getChanges (/Users/matb/projects/san/node_modules/typescript/lib/tsserver.js:136619:45)
    at Object.organizeImports (/Users/matb/projects/san/node_modules/typescript/lib/tsserver.js:129299:34)
    at Proxy.organizeImports (/Users/matb/projects/san/node_modules/typescript/lib/tsserver.js:151373:39)
    at IOSession.Session.organizeImports (/Users/matb/projects/san/node_modules/typescript/lib/tsserver.js:162453:60)
    at Session.handlers.ts.Map.ts.getEntries._a.<computed> (/Users/matb/projects/san/node_modules/typescript/lib/tsserver.js:161174:61)
    at /Users/matb/projects/san/node_modules/typescript/lib/tsserver.js:162775:88
```

### PR DESCRIPTION
**Pull request by armanio123**
Fixes #43107

Text changes will now include multiline comments if they are considered trailing comments from the previous node.

### CODE REVIEW

File: tests/cases/fourslash/organizeImports4.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+////  */
+//// import * as somethingElse from "anotherpath";
+//// import * as AnotherThing from "somepath";/** 
+////  * some comment here
+////  * and there
+////  */
+//// import * as AnotherThingElse from "someotherpath";
+
+
+verify.organizeImports('');
```
Comment (jessetrinity) [inline tests/cases/fourslash/organizeImports4.ts:17]: All of these imports are unused so they will just be deleted. We probably do want to test that but you will want to also add tests in which the imports are used, and with non-imports between the imports, also with trailing trivia.

**Review by jessetrinity — CHANGES_REQUESTED** [review]
I think you're on the right track - I'm pretty sure the overlap issue is just coming from newlines at this point.

File: src/services/textChanges.ts [hunk 2]
```diff
@@ -199,6 +201,16 @@ namespace ts.textChanges {
             // when b is deleted - we delete it
             return leadingTriviaOption === LeadingTriviaOption.IncludeAll ? fullStart : start;
         }
+
+        // if node has a trailing comments, use comment end position as the text has already been
+        // included.
+        if (hasTrailingComment) {
+            const comment = getTrailingCommentRanges(sourceFile.text, fullStart)?.[0];
+            if (comment) {
+                return comment.end;
```
[the remark this answers is not in the document]
Comment (armanio123) [reply src/services/textChanges.ts:210]: I'm not sure I understand in which condition this is an issue. Do you have a sample code I can test?

File: src/services/textChanges.ts [hunk 3]
```diff
@@ -208,25 +220,54 @@ namespace ts.textChanges {
...
+                        break;
+                    }
+
+                    // Get the end line of the comment and compare against the end line of the node.
+                    // If the comment end line position, the multiline comment expans to multiple lines
+                    // and is safe to return the end position.
+                    const commentEndLine = getLineOfLocalPosition(sourceFile, comment.end);
+                    if (commentEndLine > nodeEndLine) {
+                        const newEnd = skipTrivia(sourceFile.text, comment.end, /*stopAfterLineBreak*/ true, /*stopAtComments*/ true);
+                        return { end: newEnd, hasTrailingComment: true };
```
[the remark this answers is not in the document]
Comment (armanio123) [reply src/services/textChanges.ts:257]: Done. The new method name is a little confusing but I wasn't sure how to name it. Suggestions are welcome.

File: src/services/textChanges.ts [hunk 4]
```diff
@@ -208,25 +220,54 @@ namespace ts.textChanges {
...
+        if (trailingTriviaOption === TrailingTriviaOption.Include) {
+            // If the trailing comment is a multiline comment that extends to the next lines,
+            // return the end of the comment and track it for the next nodes to adjust.
+            const comments = getTrailingCommentRanges(sourceFile.text, end);
+            if (comments) {
+                const nodeEndLine = getLineOfLocalPosition(sourceFile, node.end);
+                for (const comment of comments) {
+                    // Single line can break the loop as trivia will only be this line.
+                    // Comments on subsequest lines are also ignored.
+                    if (comment.kind === SyntaxKind.SingleLineCommentTrivia || getLineOfLocalPosition(sourceFile, comment.pos) > nodeEndLine) {
```
[the remark this answers is not in the document]
Comment (armanio123) [reply src/services/textChanges.ts:247]: Done.

File: tests/cases/fourslash/organizeImports4.ts (hunk 1 above)
Comment (armanio123) [reply tests/cases/fourslash/organizeImports4.ts:17]: Good catch I was so focused on the issue that I didn't even considered. I did and the crash still happens. Working on it.

Comment (armanio123) [reply tests/cases/fourslash/organizeImports4.ts:17]: Fixed.

File: src/services/textChanges.ts [hunk 5]
```diff
@@ -301,13 +342,27 @@ namespace ts.textChanges {
             this.deleteRange(sourceFile, getAdjustedRange(sourceFile, node, node, options));
         }
 
+        public deleteNodes(sourceFile: SourceFile, nodes: readonly Node[], options: ConfigurableStartEnd = { leadingTriviaOption: LeadingTriviaOption.IncludeAll }): void {
+            // When deleting multiple nodes we need to track if the end position is including multiline trailing comments.
+            let hasTrailingComment = false;
+
+            for (const node of nodes) {
```
[the remark this answers is not in the document]
Comment (armanio123) [reply src/services/textChanges.ts:349]: Unfortunately, the result would be `console.log();/**` as we have no way of recognizing what the previous node is about. 

I think this is a design flaw and suggest addressing it on another ticket. So far this PR will not break that scenario.

File: src/services/textChanges.ts [hunk 6]
```diff
@@ -345,6 +405,13 @@ namespace ts.textChanges {
             this.replaceRangeWithNodes(sourceFile, getAdjustedRange(sourceFile, startNode, endNode, options), newNodes, options);
         }
 
+        public replaceAndDeleteNodeWithNodes(sourceFile: SourceFile, oldNode: Node, newNodes: readonly Node[], deleteNodes: Node[], replaceOptions: ChangeNodeOptions = useNonAdjustedPositions, deleteOptions: ConfigurableStartEnd = { leadingTriviaOption: LeadingTriviaOption.IncludeAll }): void {
```
Comment (jessetrinity) [inline src/services/textChanges.ts:408]: This name is a bit confusing - I probably wouldn't even extract this to a function. It's pretty simple and it looks like you only use it once anyway.

File: tests/cases/fourslash/organizeImports6.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
+/// <reference path="fourslash.ts" />
+
+// Regression test for GH#43107
+
+//// import * as something from "path"; /* small comment */ // single line one.
```
Comment (jessetrinity) [inline tests/cases/fourslash/organizeImports6.ts:5]: I think you want this single line comment test to also not delete the import so you can see that the comment is preserved (by adding usages of `something` and `somethingElse` below).

**Review by jessetrinity — APPROVED** [review]
Just a couple more small comments from me.

File: src/services/textChanges.ts (hunk 6 above)
Comment (armanio123) [reply src/services/textChanges.ts:408]: Agree, done.

File: tests/cases/fourslash/organizeImports6.ts (hunk 7 above)
Comment (armanio123) [reply tests/cases/fourslash/organizeImports6.ts:5]: It was actually meant to test that the single line is deleted. I found I used to have that problem. It also makes sense to test the contrary so I have updated to do it for both.

**Pull request merged**